File: kirun-js/src/engine/model/StatementGroup.ts
import { Schema } from '../json/schema/Schema';
import { SchemaType } from '../json/schema/type/SchemaType';
import { TypeUtil } from '../json/schema/type/TypeUtil';
import { Namespaces } from '../namespaces/Namespaces';
import { AbstractStatement } from './AbstractStatement';
import { Position } from './Position';

export class StatementGroup extends AbstractStatement {
    private static readonly SCHEMA_NAME: string = 'StatementGroup';
    public static readonly SCHEMA: Schema = new Schema()
        .setNamespace(Namespaces.SYSTEM)
        .setName(StatementGroup.SCHEMA_NAME)
        .setType(TypeUtil.of(SchemaType.OBJECT))
        .setProperties(
            new Map([
                ['statementGroupName', Schema.ofString('statementGroupName')],
                ['comment', Schema.ofString('comment')],
                ['description', Schema.ofString('description')],
                ['position', Position.SCHEMA],
            ]),
        );

    private statementGroupName: string;
    private statements: Map<string, boolean>;

    constructor(statementGroupName: string, statements: Map<string, boolean> = new Map()) {
        super();
        this.statementGroupName = statementGroupName;
        this.statements = statements;
    }

    public getStatementGroupName(): string {
        return this.statementGroupName;
    }
    public setStatementGroupName(statementGroupName: string): StatementGroup {
        this.statementGroupName = statementGroupName;
        return this;
    }

    public getStatements(): Map<string, boolean> {
        return this.statements;
    }

    public setStatements(statements: Map<string, boolean>): StatementGroup {
        this.statements = statements;
        return this;
    }

    public static from(json: any): StatementGroup {
        return new StatementGroup(
            json.statementGroupName,
            new Map(
                Object.entries(json.statements || {}).map(([k, v]) => [
                    k,
                    ('' + v)?.toLowerCase() == 'true',
                ]),
            ),
        )
            .setPosition(Position.from(json.position))
            .setComment(json.comment)
            .setDescription(json.description)
            .setOverride(json.override ?? false) as StatementGroup;
    }

    public toJSON(): any {
        return {
            statementGroupName: this.statementGroupName,
            statements: Object.fromEntries(this.statements),
            position: this.getPosition(),
            comment: this.getComment(),
            description: this.getDescription(),
            override: this.isOverride(),
        };
    }
}

File: kirun-py/src/kirun_py/model/statement_group.py
from __future__ import annotations
from typing import Dict, Optional

from kirun_py.model.abstract_statement import AbstractStatement
from kirun_py.model.position import Position


class StatementGroup(AbstractStatement):
    SCHEMA_NAME: str = 'StatementGroup'

    def __init__(
        self,
        statement_group_name: str,
        statements: Optional[Dict[str, bool]] = None,
    ):
        super().__init__()
        self._statement_group_name = statement_group_name
        self._statements: Dict[str, bool] = statements if statements is not None else {}

    def get_statement_group_name(self) -> str:
        return self._statement_group_name

    def set_statement_group_name(self, name: str) -> StatementGroup:
        self._statement_group_name = name
        return self

    def get_statements(self) -> Dict[str, bool]:
        return self._statements

    def set_statements(self, statements: Dict[str, bool]) -> StatementGroup:
        self._statements = statements
        return self

    @staticmethod
    def from_value(json: dict) -> StatementGroup:
        sg = StatementGroup(
            json.get('statementGroupName', ''),
            {
                k: str(v).lower() == 'true'
                for k, v in (json.get('statements') or {}).items()
            },
        )
        pos = Position.from_value(json.get('position'))
        if pos:
            sg.set_position(pos)
        if json.get('comment'):
            sg.set_comment(json['comment'])
        if json.get('description'):
            sg.set_description(json['description'])
        sg.set_override(json.get('override', False))
        return sg

    def to_json(self) -> dict:
        return {
            'statementGroupName': self._statement_group_name,
            'statements': dict(self._statements),
            'position': self.get_position().to_json() if self.get_position() else None,
            'comment': self.get_comment(),
            'description': self.get_description(),
            'override': self.is_override(),
        }
